app/gui/src/presenter/graph/state.rs

```diff
@@ -452,6 +445,23 @@ impl<'a> ControllerChange<'a> {
         }
     }
 
+    /// Set whether this node is currently awaiting completion of execution.
+    pub fn set_node_pending(
+        &self,
+        expression: ast::Id,
+        is_pending: bool,
+    ) -> Option<(ViewNodeId, bool)> {
+        let node_id = self.state.nodes.borrow().get(expression).is_some().as_some(expression)?;
+        let mut nodes = self.nodes.borrow_mut();
+        let displayed = nodes.get_mut(node_id)?;
```

Comment:
Unnecessary double borrow, access and inconsistent path. `self.state.nodes` and `self.nodes` are the same things, so I think the first line is basically unnecessary.

It looks like this first check doesn't really accomplish anything, unless the intention was to avoid `borrow_mut` on nodes when the expression ID is invalid. If that's really the case, then a comment with the reason would be very helpful. I don't recall any other code that would need to depend on this behavior. Once we are calling the setter, this borrow should be guaranteed to succeed anyway.

```suggestion
        let mut nodes = self.nodes.borrow_mut();
        let displayed = nodes.get_mut(expression)?;
```